core/src/main/java/io/grpc/internal/ServerCallImpl.java

```diff
@@ -121,33 +123,45 @@ public boolean isCancelled() {
     return cancelled;
   }
 
-  ServerStreamListenerImpl newServerStreamListener(ServerCall.Listener<ReqT> listener,
+  ServerStreamListenerImpl<ReqT> newServerStreamListener(ServerCall.Listener<ReqT> listener,
       Future<?> timeout) {
-    return new ServerStreamListenerImpl(listener, timeout);
+    return new ServerStreamListenerImpl<ReqT>(this, listener, timeout);
   }
 
   /**
    * All of these callbacks are assumed to called on an application thread, and the caller is
    * responsible for handling thrown exceptions.
    */
-  private class ServerStreamListenerImpl implements ServerStreamListener {
+  @VisibleForTesting
+  static final class ServerStreamListenerImpl<ReqT> implements ServerStreamListener {
+    private final ServerCallImpl<ReqT, ?> call;
     private final ServerCall.Listener<ReqT> listener;
     private final Future<?> timeout;
+    private boolean messageReceived;
 
-    public ServerStreamListenerImpl(ServerCall.Listener<ReqT> listener, Future<?> timeout) {
+    public ServerStreamListenerImpl(
+        ServerCallImpl<ReqT, ?> call, ServerCall.Listener<ReqT> listener, Future<?> timeout) {
+      this.call = checkNotNull(call, "call");
       this.listener = checkNotNull(listener, "listener must not be null");
-      // TODO: check if timeout should not be null
-      this.timeout = timeout;
+      this.timeout = checkNotNull(timeout, "timeout");
     }
 
     @Override
     public void messageRead(final InputStream message) {
       try {
-        if (cancelled) {
+        if (call.cancelled) {
           return;
         }
+        // Special case for unary calls.
+        if (messageReceived && call.method.getType() == MethodType.UNARY) {
+          call.close(Status.INVALID_ARGUMENT.withDescription(
```

Comment:
This isn't thread-safe because ServerCall isn't thread-safe. We could have an `internalClose` that doesn't do the `closeCalled` manipulation, but as discussed offline, let's just call `stream.close()` here.